google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineStopAction.java

```diff
@@ -0,0 +1,109 @@
+import com.intellij.execution.process.ProcessAdapter;
+import com.intellij.execution.process.ProcessEvent;
+import com.intellij.icons.AllIcons.General;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.ui.Messages;
+import com.intellij.remoteServer.runtime.deployment.DeploymentRuntime.UndeploymentTaskCallback;
+import com.intellij.remoteServer.runtime.log.LoggingHandler;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Set;
+
+import javax.swing.SwingUtilities;
+
+/**
+ * Stops an App Engine based application running on GCP
+ */
+public class AppEngineStopAction extends AppEngineAction {
+  private static final Logger logger = Logger.getInstance(AppEngineStopAction.class);
+
+  private AppEngineHelper appEngineHelper;
+  private UndeploymentTaskCallback callback;
+
+  private Set<String> modulesToStop;
+  private String versionToStop;
+
+  public AppEngineStopAction(
+      @NotNull AppEngineHelper appEngineHelper,
+      @NotNull LoggingHandler loggingHandler,
+      @NotNull Set<String> modulesToStop,
+      @NotNull String versionToStop,
+      @NotNull UndeploymentTaskCallback callback) {
+    super(loggingHandler, appEngineHelper, callback);
+
+    this.appEngineHelper = appEngineHelper;
+    this.modulesToStop = modulesToStop;
+    this.versionToStop = versionToStop;
+    this.callback = callback;
+  }
+
+  @Override
+  public void run() {
+    GeneralCommandLine commandLine = new GeneralCommandLine(
+        appEngineHelper.getGcloudCommandPath().getAbsolutePath());
+
+    commandLine.addParameters("preview", "app", "modules", "stop");
+    for(String module : modulesToStop) {
+      commandLine.addParameter(module);
+    }
+    commandLine.addParameters("--version", versionToStop);
+    try {
+      executeProcess(
+          commandLine,
+          new StopModuleProcessListener());
+    } catch (ExecutionException e) {
+      logger.error(e);
+      callback.errorOccurred(GctBundle.message("appengine.stop.modules.version.execution.error"));
+    }
+  }
+
+  private class StopModuleProcessListener extends ProcessAdapter {
+
+    @Override
+    public void processTerminated(ProcessEvent event) {
+      if (event.getExitCode() == 0) {
+        callback.succeeded();
+
+        SwingUtilities.invokeLater(new Runnable() {
+          @Override
+          public void run() {
+            Messages.showMessageDialog(
+                GctBundle.message("appengine.stop.modules.version.success.dialog.message"),
+                GctBundle.message("appengine.stop.modules.version.success.dialog.title"),
+                General.Information);
+          }
+        });
+      } else {
+        logger.error("Application stop process exited with an error. Exit Code:" + event.getExitCode());
+        callback.errorOccurred(
+            GctBundle.message("appengine.stop.modules.version.execution.error.with.code", event.getExitCode()));
+      }
+
+      deleteCredentials();
```

Comment:
I'm wondering if we want to include deleteCredentials in a finally clause to ensure it always happens.